src/blackbox.py

```diff
@@ -0,0 +1,270 @@
+        try:
+            # If a config is invalid then blackbox exporter would exit immediately. #TODO: true?
+            self._container.replan()
+        except ChangeError as e:
+            logger.error(
+                "Failed to replan; pebble plan: %s; %s",
+                self._container.get_plan().to_dict(),
+                str(e),
+            )
+
+    def update_config(self) -> None:
+        """Update blackbox exporter config file to reflect changes in configuration.
+
+        Raises:
+            ConfigUpdateFailure, if failed to update configuration file.
+        """
+        if not self.is_ready:
+            raise ContainerNotReady("cannot update config")
+        logger.debug("applying config changes")
+        config = self.model.config.get("config_file")
+        # Basic config validation: valid yaml
+        if config:
+            try:
+                yaml.safe_load(config)
+            except yaml.YAMLError as e:
+                logger.error(
+                    "Failed to load the configuration; invalid YAML: %s %s", config, str(e)
+                )
+                raise ConfigUpdateFailure("Failed to load config; invalid YAML")
+            self._container.push(self._config_path, config, make_dirs=True)
+
+    def restart_service(self) -> bool:
+        """Helper function for restarting the underlying service.
+
+        Returns:
+            True if restart succeeded; False otherwise.
+        """
+        logger.info("Restarting service %s", self._service_name)
+
+        if not self.is_ready:
+            logger.error("Cannot (re)start service: container is not ready.")
+            return False
+
+        # Check if service exists, to avoid ModelError from being raised when the service does
+        # not exist.
+        if not self._container.get_plan().services.get(self._service_name):
+            logger.error("Cannot (re)start service: service does not (yet) exist.")
+            return False
+
+        self._container.restart(self._service_name)
+
+        return True
+
+    def reload(self) -> None:
+        """Trigger a hot-reload of the configuration (or service restart).
+
+        Raises:
+            ConfigUpdateFailure, if the reload (or restart) fails.
+        """
+        if not self.is_ready:
+            raise ContainerNotReady("cannot reload")
+
+        try:
+            self.api.reload()
+        except BlackboxExporterBadResponse as e:
+            logger.warning("config reload via HTTP POST failed: %s", str(e))
+            # hot-reload failed so attempting a service restart
+            if not self.restart_service():
+                raise ConfigUpdateFailure(
+                    "Is config valid? hot reload and service restart failed."
+                )
+
+
+class BlackboxExporterBadResponse(RuntimeError):
+    """A catch-all exception type to indicate 'no reply', regardless of the reason."""
+
+
+class BlackboxExporterApi:
```

Comment:
Should we move it out into a dedicated `blackbox_exporter_client.py` file?

Reply:
I didn't really like the idea because it's only one function in that class really (to reload the configuration); also I find having two files named `blackbox.py` and `blackbox_client.py` is very unintuitive to which does what, so I'd rather keep it like this for now